### PR TITLE
puppet - replace stdout with console in logdest option

### PR DESCRIPTION
**Pull request by aminvakil**
##### SUMMARY
Fixes https://github.com/ansible-collections/community.general/issues/1190

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
puppet

##### ADDITIONAL INFORMATION
I guess it would be best if there could be an alias from `stdout` to `console` for now showing a deprecation warning, therefore users which have explicitly stated `stdout` (although it won't work) don't face issues with this change.

https://puppet.com/docs/puppet/7.6/man/agent.html
https://puppet.com/docs/puppet/6.22/man/agent.html
https://puppet.com/docs/puppet/5.5/man/agent.html

> `--logdest`
Where to send log messages. Choose between 'syslog' (the POSIX syslog service), 'eventlog' (the Windows Event Log), 'console', or the path to a log file.

### CODE REVIEW

File: plugins/modules/system/puppet.py [hunk 1]
```diff
@@ -54,10 +54,10 @@
   logdest:
     description:
     - Where the puppet logs should go, if puppet apply is being used.
-    - C(all) will go to both C(stdout) and C(syslog).
+    - C(all) will go to both C(console) and C(syslog).
     type: str
-    choices: [ all, stdout, syslog ]
-    default: stdout
+    choices: [ all, console, syslog ]
+    default: console
```
Comment (felixfontein) [inline plugins/modules/system/puppet.py:60]: You definitely have to keep the `stdout` option to avoid breaking backwards compatibility. I wonder whether the name changed over time, and originally `stdout` **was** the correct value (in which case we can't simply replace it).

Comment (aminvakil) [reply plugins/modules/system/puppet.py:60]: I have not specified it in docs, but it will be set to `console` if `stdout` has been used. Do you want to me add it to docs again, this would result in more users choosing wrong option IMO.

Comment (felixfontein) [reply plugins/modules/system/puppet.py:60]: You also have to keep it in the docs. It's probably be best to deprecate the old value by code (and mention that in the docs), this will more effectively prevent people from using it. (Also if someone is used it and it is the correct value for their version of puppet, they are more likely to complain so we find out ;) )

Comment (aminvakil) [reply plugins/modules/system/puppet.py:61]: FYI I asked this in `#puppet` IRC channel on Freenode and someone told me it's been `console` since `puppet` 2.7 at least:

from Puppet 2.7, puppet agent -help: 
```
 --logdest:
  Where to send messages. Choose between syslog, the console, and a log
  file. Defaults to sending messages to syslog, or the console if
  debugging or verbosity is enabled.
```

**Comment by aminvakil** [comment]
This is the PR which first brought up `stdout`: https://github.com/ansible/ansible/pull/42218

**Comment by felixfontein** [comment]
@aminvakil that PR only added a new choice (`all`), `stdout` was already there before. The option was added here: ansible/ansible-modules-extras#1644

File: plugins/modules/system/puppet.py (hunk 1 above)
Comment (felixfontein) [reply plugins/modules/system/puppet.py:61]: Considering that Puppet 2.7.0 is from 2011, and the option for this module was 'only' added in 2016, it should be safe to change the value. I'm wondering a bit why Puppet didn't complain about the wrong value...

File: plugins/modules/system/puppet.py [hunk 2]
```diff
@@ -54,10 +54,11 @@
   logdest:
     description:
     - Where the puppet logs should go, if puppet apply is being used.
-    - C(all) will go to both C(stdout) and C(syslog).
+    - C(all) will go to both C(console) and C(syslog).
+    - C(stdout) is deprecated and has been replaced by C(console) since 3.0.1.
```
Comment (felixfontein) [inline plugins/modules/system/puppet.py:58]:
```suggestion
    - C(stdout) is deprecated and has been replaced by C(console) since community.general 3.1.0.
```
Never mention version numbers without saying to what they belong :) Also deprecations cannot be made in bugfix releases.

File: plugins/modules/system/puppet.py [hunk 3]
```diff
@@ -268,8 +269,15 @@ def main():
         cmd = "%s apply --detailed-exitcodes " % base_cmd
         if p['logdest'] == 'syslog':
             cmd += "--logdest syslog "
+        if p['logdest'] == 'stdout':
+            cmd += "--logdest console "
+            module.warn(
```
Comment (felixfontein) [inline plugins/modules/system/puppet.py:274]: Use `module.deprecate()` for deprecation messages.

I think it's better to move the deprecation into a separate PR, and instead just mention that `stdout` will be deprecated eventually. This also gives users a better upgrade path.

Comment (aminvakil) [reply plugins/modules/system/puppet.py:274]: I was going to, but I wasn't sure what to do here, we use the word "deprecate" here over and over, when we are in fact fixing a bug. That's why I used `3.0.1` here and there, and `module.warn` instead of `module.deprecate`.

Comment (aminvakil) [reply plugins/modules/system/puppet.py]: I removed the warning, and if a user has explicitly stated `stdout` in their task, they won't be affected for now.
https://github.com/ansible-collections/community.general/pull/2407/commits/7b1428f70a4e7b5d6b4ee34f5b835ac725c0fb80

Although I'm still not sure why we can't do this all in `3.0.1` as this is a bug.

Comment (felixfontein) [reply plugins/modules/system/puppet.py]: We were mainly talking about deprecation because you wanted to remove the old value :) For a bugfix, it's best to only fix the bug and not do other changes. Strictly speaking, this PR is adding a new feature, namely a new value in `choices` for `log_dest`.

For this to be a proper bugfix that can be backported, only the code which handles `if p['logdest'] == 'all':` should be touched.

Comment (aminvakil) [reply plugins/modules/system/puppet.py]: I'd really appreciate if you could take a new look at this PR as a bugfix.
I will open a new PR after this has been merged deprecating `stdout` in c.g. 3.1.0.

File: changelogs/fragments/2407-puppet-change_stdout_to_console.yaml [hunk 4]
```diff
@@ -0,0 +1,3 @@
+---
+bugfixes:
+  - puppet - replace ``console` with ``stdout`` in ``logdest`` option when all has been chosen (https://github.com/ansible-collections/community.general/issues/1190).
```
Comment (felixfontein) [inline changelogs/fragments/2407-puppet-change_stdout_to_console.yaml:3]:
```suggestion
  - puppet - replace ``console` with ``stdout`` in ``logdest`` option when ``all`` has been chosen (https://github.com/ansible-collections/community.general/issues/1190).
```

**Review by felixfontein — COMMENTED** [review]
Looks good, except one tiny detail :)

File: plugins/modules/system/puppet.py (hunk 3 above)
Comment (felixfontein) [reply plugins/modules/system/puppet.py]: It's probably best to first mention it will be deprecated, and then deprecate it in 4.0.0. This gives users some time where the old value still works without triggering deprecation warnings.

**Review by felixfontein — APPROVED** [review]

**Comment by felixfontein** [comment]
@aminvakil thanks for fixing this!
